pkg/segment/query/processor/queryprocessor.go

```diff
@@ -80,12 +80,11 @@ func NewQueryProcessor(firstAgg *structs.QueryAggregators, queryInfo *query.Quer
 		if dataProcessor == nil {
 			break
 		}
-		dataProcessor.qid = searcher.qid
 		dataProcessors = append(dataProcessors, dataProcessor)
 	}
 
 	// Hook up the streams (searcher -> dataProcessors[0] -> ... -> dataProcessors[n-1]).
-	if len(dataProcessors) > 0 && !dataProcessors[0].IsDataGenerator() {
+	if len(dataProcessors) > 0 && !dataProcessors[0].CheckAndSetQidForDataGenerator(searcher.qid, true) {
 		dataProcessors[0].streams = append(dataProcessors[0].streams, NewCachedStream(searcher))
 	}
```

Comment:
nit: I think this change makes the code less clear. `CheckAndSetQidForDataGenerator` is probably doing too much. You might want to keep the check for `IsDataGenerator()` and then only call `CheckAndSetQidForDataGenerator` if necessary.

You can ignore this comment if you think doing it the way you are is the best approach

Reply:
addressed in the latest commit.